third_party/intel/lib/TritonIntelGPUToLLVM/TargetInfo.cpp

```diff
@@ -107,6 +108,77 @@ Value TargetInfo::programId(RewriterBase &rewriter, Location loc,
+                         Value acc, unsigned numLanesToReduce,
+                         unsigned warpSize) {
+  auto spvGroupOp = spirv::GroupOperation::Reduce;
+  Value clusterSize;
+  if (numLanesToReduce != warpSize) {
+    spvGroupOp = spirv::GroupOperation::ClusteredReduce;
+    clusterSize = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getI32Type(),
+        rewriter.getI32IntegerAttr(numLanesToReduce));
+  }
+
+  Value result = rewriter.create<GroupOp>(loc, resultTy, spirv::Scope::Subgroup,
+                                          spvGroupOp, acc, clusterSize);
+  return result;
+}
+
+Value warpReduceHelper(RewriterBase &rewriter, Location loc, Value acc,
+                       Operation *reduceOp, unsigned numLanesToReduce,
+                       unsigned warpSize) {
+  auto resultType = reduceOp->getResult(0).getType();
+  Value warpReduce =
+      llvm::TypeSwitch<mlir::Operation *, Value>(reduceOp)
+          .Case<arith::AddFOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformFAddOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::AddIOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformIAddOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::MulFOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformFMulOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::MulIOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformIMulOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::MaxNumFOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformFMaxOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::MinNumFOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformFMinOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::AndIOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformBitwiseAndOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::OrIOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformBitwiseOrOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Case<arith::XOrIOp>([&](auto) {
+            return createSPIRVGroupOp<spirv::GroupNonUniformBitwiseXorOp>(
+                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
+          })
+          .Default([](auto) {
+            llvm_unreachable("Unsupported reduction");
+            return Value();
+          });
```

Comment:
```suggestion
template <typename OpTy>
struct spirv_group_op {};

template <>
struct spirv_group_op<arith::AddFOp> {
  using type = spirv::GroupNonUniformFAddOp;
};
// ...
template <typename OpTy>
using spirv_group_op_ty = spirv_group_op<OpTy>::type;
// ----
  Value warpReduce =
      TypeSwitch<mlir::Operation *, Value>(reduceOp)
          .Case<arith::AddFOp, arith::AddIOp, arith::MulFOp, arith::MulIOp, arith::MaxNumFOp, arith::MinNumFOp, arith::AndIOp, arith::OrIOp, arith::XOrIOp>([&](auto reduceOp) {
            return createSPIRVGroupOp<spirv_group_op_ty<decltype(reduceOp)>>(
                rewriter, loc, resultType, acc, numLanesToReduce, warpSize);
          });
```
No need for that default as this is implicit.

Struct type approach looks cleaner to me.

`TypeSwitch` is defined under `::mlir` as an alias